README
##[_--{SMAUG II}--_] (http://smaugmuds.github.io/_smaug_)

###Introduction

```
                    R E A L M S    O F    D E S P A I R  !
  ___________________________________________________________________________
 //            /                                                            \\
[|_____________\   ********   *        *   ********   *        *   *******   |]
[|   \\._.//   /  **********  **      **  **********  **      **  *********  |]
[|   (0...0)   \  **********  ***    ***  **********  ***    ***  *********  |]
[|    ).:.(    /  ***         ****  ****  ***    ***  ***    ***  ***        |]
[|    {o o}    \  *********   **********  **********  ***    ***  *** ****   |]
[|   / ' ' \   /   *********  *** ** ***  **********  ***    ***  ***  ****  |]
[|-'- /   \ -`-\         ***  ***    ***  ***    ***  ***    ***  ***   ***  |]
[|   .VxvxV.   /   *********  ***    ***  ***    ***  **********  *********  |]
[|_____________\  **********  **      **  **      **  **********  *********  |]
[|             /  *********   *        *  *        *   ********    *******   |]
 \\____________\____________________________________________________________//
    |                                                                     |
    |    --{ [S]imulated [M]edieval [A]dventure Multi[U]ser [G]ame }--    |
    |_____________________________________________________________________|
    |                                                                     |
    |                              README                                 |
    |_____________________________________________________________________|
    |                                                                     |
    |                          --{SMAUG II}--                             |
    |_____________________________________________________________________|
   //                                                                     \\
  [|  SMAUG 2.0 © 2014-2015 Antonio Cao (@burzumishi)                      |]
  [|                                                                       |]
  [|  AFKMud Copyright 1997-2007 by Roger Libiez (Samson),                 |]
  [|  Levi Beckerson (Whir), Michael Ward (Tarl), Erik Wolfe (Dwip),       |]
  [|  Cameron Carroll (Cam), Cyberfox, Karangi, Rathian, Raine,            |]
  [|  Xorith, and Adjani.                                                  |]
  [|  All Rights Reserved. External contributions from Remcon, Quixadhal,  |]
  [|  Zarius and many others.                                              |]
  [|                                                                       |]
  [|  SMAUG 1.4 © 1994-1998 Thoric/Altrag/Blodkai/Narn/Haus/Scryn/Rennard  |]
  [|  Swordbearer/Gorog/Grishnakh/Nivek/Tricops/Fireblade/Edmond/Conran    |]
  [|                                                                       |]
  [|  Merc 2.1 Diku Mud improvments © 1992-1993 Michael Chastain, Michael  |]
  [|  Quan, and Mitchell Tse. Original Diku Mud © 1990-1991 by Sebastian   |]
  [|  Hammer, Michael Seifert, Hans Henrik St{rfeldt, Tom Madsen, Katja    |]
  [|  Nyboe. Win32 port Nick Gammon.                                       |]
   \\_____________________________________________________________________//

```

**--{ [S]imulated [M]edieval [A]dventure Multi-[U]ser [G]ame }--**

**--{SMAUG II}--** is a **Merc** and **DikuMUD** derived MUD server. Its name is a backronym inspired by the dragon **Smaug** found in **J. R. R. Tolkien's** fiction. **--{SMAUG II}--** is a multiplayer text-based role-playing game in a real-time virtual world, which is a type of MUD. Modern massively multiplayer online role-playing games **(MMORPGs)** traces directly back to the MUD genre.

**--{SMAUG II}--** is an enhancenment of **SMAUG 1.8b**, this project was started to have a source code structure which builds, installs and runs in most modern distributed system environments. **--{SMAUG II}--** is ready to add, enable and disable new code snippets using the **GNU Autotools** build system. Many bug fixes and additions were made to the **SMAUG 1.8b** source code, always trying to keep the compatibility with **SMAUG 1.8b**. All paths for data files and directories were modified, because **--{SMAUG II}--** installation is stored into a **LSB** directory distribution. All code is internationalized using **GNU Gettext**, all zones, clases, races and some system files were modified to be translated. Old "**startd**" script was replaced with "**smaugd**" and "**smaugd-monitor**", these scripts will start up and keep your MUD server running. If **--{SMAUG II}--** server process fails, "**smaugd-monitor**" will respawn it again. To know more about all these new changes, take a look at the section **--{SMAUG II}-- ChangeLog**. Some new code snippets were extracted and adapted from **AFKMud** and **SMAUGfuss**. All licenses and mentions to the original **SMAUG** code authors are displayed in source code header files.

**SMAUG 1.8b**, is an fixed **SMAUG 1.4a** which is a heavily modified derivative of the **Merc 2.1** distribution. **SMAUG 1.4a** is close to six times the size of **Merc 2.1**, and has major changes to the inner workings of the code, hence making porting of code from other Diku-derivatives non-trivial to the non-coder. Due to the ability to edit most everything online, you will most likely find the online help pages more informative. **SMAUG** is the base code that runs the popular **Realms of Despair** MUD. Through the years it has been tweaked, modified, and enhanced to become one of the most feature packed branches of the **MERC** family of codebases. 

**SMAUG** is a derivative of **Merc Diku Mud** is a **Diku Mud** with many enhancements and contributions. It was written in 1990 and 1991 by **Sebastian Hammer**, **Tom Madsen**, **Katja Nyboe**, **Michael Seifert**, and **Hans Henrik StÃŠrfeldt** at **DIKU** _(Datalogisk Institut KÃžbenhavns Universitet) at the department of computer science at the University of Copenhagen in Copenhagen, Denmark_. Commonly referred to as simply **"Diku"**, the game was greatly inspired by **AberMUD**, though **Diku** became one of the first multi-usergames to become popular as a freely-available program for its gameplay and similarity to **Dungeons & Dragons**.

**AberMUD** was the first popular open source MUD. It was named after the town _Aberystwyth_, in which it was written. The first version was written in **B** by **Alan Cox**, **Richard Acott**, **Jim Finnis**,and **Leon Thrane** based at **University of Wales**, _Aberystwyth_ for an old **Honeywell** mainframe and opened in 1987.


### **--{SMAUG II}--** Releases

**--{SMAUG II}--** version **2.0** is HERE! With many bugfixes as well as important and useful new code and build features.

 * **[--{SMAUG II}-- Release] (https://github.com/smaugmuds/_smaug_/releases/tag/2.0) ¡Launch!**


#### Contents of the Release

	- src:                Source files.
	- bin:	              Init Scripts.
	- etc:                smaugd & httpd Script Config Files.
	- etc/init.d:         smaugd & httpd Init Scripts.
	- doc:                Documentation.
	- man:                Man Pages.
	- data/area/LC_ALL:   Realm Area files.
	- data/backup:        Backup Player files (initially empty).
	- data/boards:        Boards files.
	- data/building:      Building System files.
	- data/clans:         Clans files.
	- data/classes:       Classes files.
	- data/councils:      Councils files.
	- data/corpses:       Death corpses (initially empty).
	- data/deity:	    	  Realm Deities.
	- data/deleted:       Deleted Player files (initially empty).
	- data/houses:        Houses files.
	- data/gods:          God players.
	- data/mudprogs:      MUD Program files.
	- data/player:        Player files (initially empty).
	- data/system/LC_ALL: MUD Game System files.
	- data/races:	    	  Races files.
	- data/vault:	    	  Guilds Vaults
	- data/web:	    	    Web files & Online Client


#### Features

Here are just some of the features **--{SMAUG II}--** has:

	- Full featured, Command-Line Online Building (including shops, resets, repairshops, etc)
	- Optional, easy to use menu-based online building
	- Online building updated to Oasis OLCII
	- New Online Building features and Buildwalk
	- Damageable and repairable equipment
	- MUDPrograms/scripts (mob, obj and room programs)
	- Monsters that will remember you, hunt you down, and summon you
	- Levers, switches, pullchains, buttons and traps
	- Searchable hidden objects and doors
	- More flexibility to command syntax
	- Delayed teleport rooms, tunnels
	- Clans, pkilling, guilds, orders and councils
	- Revamped experience and skill/spell learning system
	- Unlimited bulletin boards that can be added online
	- Private mail and comments on players
	- Automatic daily scheduled rebooting
	- Advanced immortal heirarchy
	- Complicated monsters with stats, attacks, defenses and bodyparts
	- Resistances, Immunities and Susceptibilities
	- More spells, more skills, more classes, more races
	- Weapon proficiencies
	- Pipes and herb smoking
	- Full featured container commands... fill, empty, etc
	- Extra exits - ne, nw, se, sw and more -
	- Special exit flags, floorless rooms (falling)
	- More room, mob and object flags, class restricted items
	- Languages, improved string hashing code
	- All kinds of 'spelled equipment' support
	- A new flexible spell system that allows online creation and editing
	- Object grouping, items hidden under other items, layered clothing
	- The ability to dig up and bury items
	- Player corpses are saved in case of crash or reboot
	- Stable code with hundreds of bug fixes
	- A fully configurable deity system
	- Many new admin commands such as an extensive ban, immhost ipcompare and more
	- Polymorph code that can be added, edited and deleted online
	- Different prompts when you are fighting and when you are not
	- Pets saving with their owners
	- Configurable color codes online
	- Projectiles are fully supported
	- Extended bit vectors for when 32 are just not enough
	- SHA256 player passwords encryption.
	- New personalized Colors.
	- Room Mapper.
	- New Weather.
	- Timezone.
	- Holidays.
	- Server and Characters Translations.
	- Room Planes
	- Mounts and much much more
	- Fixed Liquids & Mixtures.
  - Gold, Silver and Copper based economy.
  - Gold loot rate by race.
	- Banking System.
  - Undertaker.
	- Questmasters.
  - New Classes & Races.
	- Player marriage system.
	- A Dice Game.
	- Mongoose web server with jta26 as Online MUD Web Client.

Realm Areas:
```
    +---------------------------+-------------+
    |            Area           |    Level    |
    +---------------------------+-------------+
    |   Spectral Gate           |    0 - 60   |
    |   New Darkhaven           |    0 - 60   |
    |   Darkhaven Academy       |    1 - 5    |
    |   Dwarven Daycare         |    1 - 5    |
    |   Pixie Forest            |    1 - 5    |
    |   Tullfuhrzky Manor       |    0 - 10   |
    |   Graveyard               |    5 - 10   |
    |   Haon Dor                |    5 - 10   |
    |   Miden'nir               |    5 - 15   |
    |   Unholy Grounds          |    5 - 15   |
    |   Holy Grove              |    5 - 20   |
    |   Shattered Refuge        |    5 - 20   |
    |   Dragon Exports          |    5 - 25   |
    |   Sewer                   |    5 - 30   |
    |   Dwarven Village         |   10 - 35   |
    |   Darkhaven Art Gallery   |   10 - 50   |
    |   Chapel Catacombs        |   15 - 25   |
    |   The Astral Plane        |   15 - 35   |
    |   Redferne's Residence    |   20 - 30   |
    |   Build Interface Stuff   |    0 - 60   |
    |   Gods                    |    0 - 60   |
    |   Limbo                   |    0 - 60   |
    +---------------------------+-------------+
```


#### **--{SMAUG II}--** ChangeLog

	* Changes from **SMAUG 1.8b** to **--{SMAUG II}--**:
```
		- Moved all "KEY" definitions to "mud.h".
		- Added Quest Masters.
		- Added new Classes & Races.
		- Added Customizable Command Aliases.
		- Added Arena Challenge.
		- Added Weapon Proficency snippet and Archery.
		- Added Gold, Silver and Copper based economy.
		- Enabled Gold loot multirate by race.
		- Added Undertaker snippet (ACT_UNDERTAKER).
		- Added "Banking" code snippet.
		- Added "Marriage" code snippet.
		- Added a Dice game.
		- Added Oasis OLCII Online editor, new features and "do_buildwalk".
		- Defined new MUD default colors.
		- Adapted Room Mapper, Colors, Weather, Timezone, Holidays systems from AFKMud.
		- Added Mongoose web server with JTA26.jar MUD WebClient.
		- Added GNU Gettext support. Translated: Server, Areas, Races, Classes, Commands, Skills & Socials.
		- Required runtime data directories moved to "data".
		- Renamed "services.c" to "win32srv.c".
		- Renamed "comm.c" to "smaug.c".
		- Removed all InterMud code: "IMC2" & "ICE(D)". If you want Intermud support maybe you can implement I3 or any other Intermud Chat source snippet.
		- Fixed "Liquids & Mixtures" system, replaces old "do_drink", "do_empty" & "do_fill" from "misc.c" routines.
		- Fixed many bugs and redundant definitions.
		- Indented code with GNU style.
		- Init Script Daemon, Monitor and Configuration files: "smaugd.conf" & "smaugd" & "smaugd-monitor".
		- Updated documentation files.
		- Added "Man Page" files for "smaug" and "smaugd".
		- Conversion to "GNU Autotools" build system.
		- Added "PREFIX", "DESTDIR" & "RUNDIR" to "configure.ac", "smaug.h" and "mud.h".
		- "LOG_DIR" variable in "mud.h" is now obsolete and has been changed to "LOGDIR" in "smaug.h", all server log files will be stored to "$(localstatedir)/log/smaug/" system directory.
		- Installed files are now splitted in several libraries and binaries using "Ranlib & Libtool": "libsmaug-act.so", "libsmaug-player.so", "libsmaug-mudprog.so", "libsmaug-info.so", "libsmaug-realm.so", "smaug", "smaugd" & "smaugd-monitor".
		- Configure checks for "cc", "gcc", "g++", "(g)awk", "-ldl", "-lz", "-lcrypt", "-lssl", "basic C# libraries", "C# headers & functions", "bash", "env", "perl".
		- Configure options to include required definitions for enabling or disabling code snippets and flags:

			Configured --{SMAUG}-- build flags:

			 * Time Format fix:     <no>
			 * RE_EXEC fix:         <yes>
			 * Profiling flags:     <yes>
			 * Developer flags:     <no>
			 * Performance flags:   <yes>

			Configured --{SMAUG}-- MUD Server features:

			 * SMAUG 2.0:           <yes>
			 * Oasis OLC:           <yes>
			 * Oasis OLC Extras:    <yes>
			 * Buildwalk:           <yes>
			 * Billion MAX_VNUM:    <yes>
			 * Password Encryption: <yes>
			 * Request Support:     <no>
			 * Room Planes:         <yes>
			 * Who Arguments:       <no>
			 * MUD HotBoot:         <yes>
			 * New Score Tables:    <yes>
			 * Show Stance Info:    <yes>
			 * Player Bleeding:     <yes>
			 * New Colors:          <yes>
			 * New MUD Mapper:      <yes>
			 * MSSP Support:        <yes>
			 * New Weather:         <yes>
			 * Timezone:            <yes>
			 * Holidays:            <yes>
			 * Honour Code:         <yes>
			 * Player Houses:       <yes>
			 * House Mobs:          <yes>
			 * Extended Rooms:      <yes>
			 * Liquids System:      <yes>
			 * Act Comm Scramble:   <no>
			 * Undertaker:          <yes>
			 * Gold Silver Copper:  <yes>
			 * Gold Race Rate:      <yes>
			 * Bank:                <yes>
			 * Player Marry:        <yes>
			 * Dice Game:           <yes>
			 * DNS Resolver:        <yes>
			 * Web Server:          <yes>

```

#### TODO

This is a list of TODO stuff:

	- New weather system.
	- Implement MXP, MSSP & MCCP protocols.
	- More useful code snippets.

#### Supported Operating Systems

	- Linux
	- BSD
	- Solaris

po/ChangeLog
2015-08-31  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.19.4.

2015-08-31  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.19.4.

2015-07-17  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.19.4.

2015-07-17  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.19.4.

2015-07-16  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.19.4.

2015-07-16  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.19.4.

2015-07-16  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.19.4.

2015-07-16  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.19.4.

2015-07-16  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.19.4.

2015-07-13  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.19.4.
	* Rules-quot: Upgrade to gettext-0.19.4.

2015-05-17  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.18.1.

2015-05-16  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.18.1.

2015-05-16  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.18.1.

2015-05-15  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.18.1.

2015-05-14  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.18.1.

2015-05-12  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.18.1.

2015-05-12  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.18.1.

2015-05-11  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.18.1.

2015-05-11  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.18.1.

2015-05-11  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.18.1.

2015-05-11  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.18.1.

2015-05-11  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.18.1.

2015-05-11  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.18.1.

2015-05-11  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.18.1.

2015-05-10  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.18.1.

2015-05-10  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.18.1.

2015-05-06  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.18.1.

2015-05-06  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.18.1.

2015-05-06  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.18.1.

2015-05-05  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.18.1.

2015-05-05  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.18.1.
	* Rules-quot: Upgrade to gettext-0.18.1.

2015-05-07  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.19.4.

2015-05-07  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.19.4.

2015-05-07  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.19.4.

2015-05-06  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.19.3.

2015-05-06  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.19.3.

2015-05-05  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.19.3.

2015-05-04  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.19.3.

2015-05-04  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.19.3.

2015-05-04  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.19.3.

2015-05-04  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.19.3.

2015-05-04  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.19.3.

2015-05-04  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.19.3.

2015-05-04  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.19.3.

2015-05-03  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.19.3.
	* Rules-quot: Upgrade to gettext-0.19.3.

2015-05-03  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.18.1.

2015-05-03  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.18.1.

2015-05-02  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.18.1.

2015-05-02  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.18.1.

2015-05-02  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.18.1.

2015-05-01  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.18.1.

2015-05-01  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.18.1.

2015-05-01  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.18.1.

2015-05-01  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.18.1.

2015-05-01  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.18.1.

2015-04-30  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.18.1.
	* Rules-quot: Upgrade to gettext-0.18.1.

2015-04-30  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.in.in: New file, from gettext-0.19.3.


src/dns.c
/*
                     R E A L M S    O F    D E S P A I R  !
   ___________________________________________________________________________
  //            /                                                            \\
 [|_____________\   ********   *        *   ********   *        *   *******   |]
 [|   \\._.//   /  **********  **      **  **********  **      **  *********  |]
 [|   (0...0)   \  **********  ***    ***  **********  ***    ***  *********  |]
 [|    ).:.(    /  ***         ****  ****  ***    ***  ***    ***  ***        |]
 [|    {o o}    \  *********   **********  **********  ***    ***  *** ****   |]
 [|   / ' ' \   /   *********  *** ** ***  **********  ***    ***  ***  ****  |]
 [|-'- /   \ -`-\         ***  ***    ***  ***    ***  ***    ***  ***   ***  |]
 [|   .VxvxV.   /   *********  ***    ***  ***    ***  **********  *********  |]
 [|_____________\  **********  **      **  **      **  **********  *********  |]
 [|             /  *********   *        *  *        *   ********    *******   |]
  \\____________\____________________________________________________________//
     |                                                                     |
     |    --{ [S]imulated [M]edieval [A]dventure Multi[U]ser [G]ame }--    |
     |_____________________________________________________________________|
     |                                                                     |
     |                     -*- DNS Resolver Module -*-                     |
     |_____________________________________________________________________|
    //                                                                     \\
   [|  SMAUG 2.0 © 2014-2015 Antonio Cao (@burzumishi)                      |]
   [|                                                                       |]
   [|  AFKMud Copyright 1997-2007 by Roger Libiez (Samson),                 |]
   [|  Levi Beckerson (Whir), Michael Ward (Tarl), Erik Wolfe (Dwip),       |]
   [|  Cameron Carroll (Cam), Cyberfox, Karangi, Rathian, Raine,            |]
   [|  Xorith, and Adjani.                                                  |]
   [|  All Rights Reserved. External contributions from Remcon, Quixadhal,  |]
   [|  Zarius and many others.                                              |]
   [|                                                                       |]
   [|  SMAUG 1.4 © 1994-1998 Thoric/Altrag/Blodkai/Narn/Haus/Scryn/Rennard  |]
   [|  Swordbearer/Gorog/Grishnakh/Nivek/Tricops/Fireblade/Edmond/Conran    |]
   [|                                                                       |]
   [|  Merc 2.1 Diku Mud improvments © 1992-1993 Michael Chastain, Michael  |]
   [|  Quan, and Mitchell Tse. Original Diku Mud © 1990-1991 by Sebastian   |]
   [|  Hammer, Michael Seifert, Hans Henrik St{rfeldt, Tom Madsen, Katja    |]
   [|  Nyboe. Win32 port Nick Gammon.                                       |]
    \\_____________________________________________________________________//
*/

#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <unistd.h>
#include <errno.h>

#include "mud.h"

DNS_DATA *first_cache;
DNS_DATA *last_cache;

void save_dns( void );

void prune_dns( void )
{
   DNS_DATA *cache, *cache_next;

   for( cache = first_cache; cache; cache = cache_next )
   {
      cache_next = cache->next;

      /*
       * Stay in cache for 14 days 
       */
      if( current_time - cache->time >= 1209600 || !str_cmp( cache->ip, "Unknown??" )
          || !str_cmp( cache->name, "Unknown??" ) )
      {
         STRFREE( cache->ip );
         STRFREE( cache->name );
         UNLINK( cache, first_cache, last_cache, next, prev );
         DISPOSE( cache );
      }
   }
   save_dns(  );
}

void check_dns( void )
{
   if( current_time >= new_boot_time_t )
      prune_dns(  );
}

void add_dns( char *dhost, char *address )
{
   DNS_DATA *cache;

   CREATE( cache, DNS_DATA, 1 );
   cache->ip = STRALLOC( dhost );
   cache->name = STRALLOC( address );
   cache->time = current_time;
   LINK( cache, first_cache, last_cache, next, prev );

   save_dns(  );
}

char *in_dns_cache( char *ip )
{
   DNS_DATA *cache;
   static char dnsbuf[MAX_STRING_LENGTH];

   dnsbuf[0] = '\0';

   for( cache = first_cache; cache; cache = cache->next )
   {
      if( !str_cmp( ip, cache->ip ) )
      {
         strncpy( dnsbuf, cache->name, MAX_STRING_LENGTH );
         break;
      }
   }
   return dnsbuf;
}

void fread_dns( DNS_DATA * cache, FILE * fp )
{
   const char *word;
   bool fMatch;

   for( ;; )
   {
      word = feof( fp ) ? "End" : fread_word( fp );
      fMatch = FALSE;

      switch ( UPPER( word[0] ) )
      {
         case '*':
            fMatch = TRUE;
            fread_to_eol( fp );
            break;

         case 'E':
            if( !str_cmp( word, "End" ) )
            {
               if( !cache->ip )
                  cache->ip = STRALLOC( "Unknown??" );
               if( !cache->name )
                  cache->name = STRALLOC( "Unknown??" );
               return;
            }
            break;

         case 'I':
            KEY( "IP", cache->ip, fread_string( fp ) );
            break;

         case 'N':
            KEY( "Name", cache->name, fread_string( fp ) );
            break;

         case 'T':
            KEY( "Time", cache->time, fread_number( fp ) );
            break;
      }

      if( !fMatch )
         bug( "fread_dns: no match: %s", word );
   }
}

void load_dns( void )
{
   char filename[256];
   DNS_DATA *cache;
   FILE *fp;

   first_cache = NULL;
   last_cache = NULL;

   snprintf( filename, 256, "%s", DNS_FILE );

   if( ( fp = fopen( filename, "r" ) ) != NULL )
   {
      for( ;; )
      {
         char letter;
         char *word;

         letter = fread_letter( fp );
         if( letter == '*' )
         {
            fread_to_eol( fp );
            continue;
         }

         if( letter != '#' )
         {
            bug( "%s: # not found.", __FUNCTION__ );
            break;
         }

         word = fread_word( fp );
         if( !str_cmp( word, "CACHE" ) )
         {
            CREATE( cache, DNS_DATA, 1 );
            fread_dns( cache, fp );
            LINK( cache, first_cache, last_cache, next, prev );
            continue;
         }
         else if( !str_cmp( word, "END" ) )
            break;
         else
         {
            bug( "%s: bad section: %s.", __FUNCTION__, word );
            continue;
         }
      }
      fclose( fp );
      fp = NULL;
   }
   prune_dns(  ); /* Clean out entries beyond 14 days */
}

void save_dns( void )
{
   DNS_DATA *cache;
   FILE *fp;
   char filename[256];

   snprintf( filename, 256, "%s", DNS_FILE );

   if( !( fp = fopen( filename, "w" ) ) )
   {
      bug( "%s: fopen", __FUNCTION__ );
      perror( filename );
   }
   else
   {
      for( cache = first_cache; cache; cache = cache->next )
      {
         fprintf( fp, "#CACHE\n" );
         fprintf( fp, "IP		%s~\n", cache->ip );
         fprintf( fp, "Name		%s~\n", cache->name );
         fprintf( fp, "Time		%ld\n", cache->time );
         fprintf( fp, "End\n\n" );
      }
      fprintf( fp, "#END\n" );
      fclose( fp );
      fp = NULL;
   }
}

/* DNS Resolver code by Trax of Forever's End */
/*
 * Almost the same as read_from_buffer...
 */
bool read_from_dns( int fd, char *buffer )
{
   static char inbuf[MAX_STRING_LENGTH * 2];
   unsigned int iStart;
   int i, j, k;

   /*
    * Check for overflow. 
    */
   iStart = strlen( inbuf );
   if( iStart >= sizeof( inbuf ) - 10 )
   {
      bug( "%s: DNS input overflow!!!", __FUNCTION__ );
      return FALSE;
   }

   /*
    * Snarf input. 
    */
   for( ;; )
   {
      int nRead;

      nRead = read( fd, inbuf + iStart, sizeof( inbuf ) - 10 - iStart );
      if( nRead > 0 )
      {
         iStart += nRead;
         if( inbuf[iStart - 2] == '\n' || inbuf[iStart - 2] == '\r' )
            break;
      }
      else if( nRead == 0 )
      {
         return FALSE;
      }
      else if( errno == EWOULDBLOCK )
         break;
      else
      {
         perror( "Read_from_dns" );
         return FALSE;
      }
   }

   inbuf[iStart] = '\0';

   /*
    * Look for at least one new line.
    */
   for( i = 0; inbuf[i] != '\n' && inbuf[i] != '\r'; i++ )
   {
      if( inbuf[i] == '\0' )
         return FALSE;
   }

   /*
    * Canonical input processing.
    */
   for( i = 0, k = 0; inbuf[i] != '\n' && inbuf[i] != '\r'; i++ )
   {
      if( inbuf[i] == '\b' && k > 0 )
         --k;
      else if( isascii( inbuf[i] ) && isprint( inbuf[i] ) )
         buffer[k++] = inbuf[i];
   }

   /*
    * Finish off the line.
    */
   if( k == 0 )
      buffer[k++] = ' ';
   buffer[k] = '\0';

   /*
    * Shift the input buffer.
    */
   while( inbuf[i] == '\n' || inbuf[i] == '\r' )
      i++;
   for( j = 0; ( inbuf[j] = inbuf[i + j] ) != '\0'; j++ )
      ;

   return TRUE;
}

/* DNS Resolver code by Trax of Forever's End */
/*
 * Process input that we got from resolve_dns.
 */
void process_dns( DESCRIPTOR_DATA * d )
{
   char address[MAX_INPUT_LENGTH];

   address[0] = '\0';

   if( !read_from_dns( d->ifd, address ) || address[0] == '\0' )
      return;

   if( address[0] != '\0' )
   {
      add_dns( d->host, address );  /* Add entry to DNS cache */
      STRFREE( d->host );
      d->host = STRALLOC( address );
   }

   /*
    * close descriptor and kill dns process 
    */
   if( d->ifd != -1 )
   {
      close( d->ifd );
      d->ifd = -1;
   }
}

/* DNS Resolver hook. Code written by Trax of Forever's End */
void resolve_dns( DESCRIPTOR_DATA * d, long ip )
{
   int fds[2];
   pid_t pid;

   /*
    * create pipe first 
    */
   if( pipe( fds ) != 0 )
   {
      perror( "resolve_dns: pipe: " );
      return;
   }

   if( dup2( fds[1], STDOUT_FILENO ) != STDOUT_FILENO )
   {
      perror( "resolve_dns: dup2(stdout): " );
      return;
   }

   if( ( pid = fork(  ) ) > 0 )
   {
      /*
       * parent process 
       */
      d->ifd = fds[0];
      d->ipid = pid;
      close( fds[1] );
   }
   else if( pid == 0 )
   {
      /*
       * child process 
       */
      char str_ip[64];
      int i;

      d->ifd = fds[0];
      d->ipid = pid;

      for( i = 2; i < 255; ++i )
         close( i );

      snprintf( str_ip, 64, "%ld", ip );
      execl( "./smaug-resolv", "DNS Resolver", str_ip, NULL );
      /*
       * Still here --> hmm. An error. 
       */
      bug( "%s: Exec failed; Closing child.", __FUNCTION__ );
      d->ifd = -1;
      d->ipid = -1;
      exit( 0 );
   }
   else
   {
      /*
       * error 
       */
      perror( "resolve_dns: failed fork" );
      close( fds[0] );
      close( fds[1] );
   }
}

void do_cache( CHAR_DATA * ch, char *argument )
{
   DNS_DATA *cache;
   int ip = 0;

   send_to_pager( "&YCached DNS Information\r\n", ch );
   send_to_pager( "IP               | Address\r\n", ch );
   send_to_pager( "------------------------------------------------------------------------------\r\n", ch );
   for( cache = first_cache; cache; cache = cache->next )
   {
      pager_printf( ch, "&W%16.16s  &Y%s\r\n", cache->ip, cache->name );
      ip++;
   }
   pager_printf( ch, "\r\n&W%d IPs in the cache.\r\n", ip );
}
